### ISSUE TITLE
resource_timeout doesn't work in qt5

### ISSUE THREAD

**Issue by kmike** [scrapinghub/splash#269]
QT5 starts printing "device not open" messages in an infinite loop when request is aborted in the middle of download.

I've opened an issue in qt bug tracker: https://bugreports.qt.io/browse/QTBUG-47654.
A short script to reproduce it: https://gist.github.com/kmike/ff287998e02fa953b4a2.
This issue makes tests hang on Travis (https://github.com/scrapinghub/splash/pull/260).

It looks like a blocker for qt5 branch merging; I'm not sure `resource_timeout` is the only condition which can trigger this bug (what about a regular timeout?).

**Comment by kmike** [comment]
It worths checking how phantomjs2 handles this.

**Comment by sunu** [comment]
phantomjs bundles their own copy of Qt, so they have their own timeout handler. See https://github.com/ariya/phantomjs/commit/52883ced681854f453029822f466b6a4a18c0e3e

**Comment by kmike** [comment]
Right, but does it work for pages like in the example?

**Comment by sunu** [comment]
Haven't tested yet. But if we disconnect the `finished` signal from the reply before calling abort() at https://github.com/scrapinghub/splash/commit/4814ad8383f4aa94cb8b962af7781fc24bc49502#diff-a8626db30f06ca48dc15a1223b9c7d1bR159, it may just work.

**Comment by sunu** [comment]
Hmm .. I tried `reply.disconnect()` which probably disconnects all the signals and the infinite loop was no more there and I got 504, 'Timeout exceeded rendering page".  But that means the whole page will time out if any single resource times out.

**Comment by kmike** [comment]
The problem is that QWebKit connects its own finished signals, so the issue is not fixed if we disconnect only our `finished` handlers. In the gist example there is no manually connected `finished` handlers. It doesn't go to infinite loop if we disconnect all handlers (like this: https://gist.github.com/kmike/ff287998e02fa953b4a2#file-x-py-L35), but it is not a solution because the reply is not handled correctly in this case.

**Comment by sunu** [comment]
I was actually talking about disconnecting Qt's finished signal. Something like http://stackoverflow.com/questions/13207493/qnetworkreply-and-qnetworkaccessmanager-timeout-in-http-request

**Comment by kmike** [comment]
@sunu for me it is not working because `finished` signal handlers are connected by QT internally. Or maybe I don't understand your point.

**Comment by sunu** [comment]
@kmike https://gist.github.com/sunu/e3c65b6b411b45390b6b - diff and test result. 

Although according to http://pyqt.sourceforge.net/Docs/PyQt5/pyqt4_differences.html the old style signals are not supported now and `disconnect` desconnects all connections and doesn't take any arguments. So there is no way of disconnecting individual signals. :(

**Comment by kmike** [comment]
@sunu we can't disconnect individual QT-installed signal handlers anyways because they are connected deep in C++ code.

**Comment by kmike** [comment]
I wasn't able to find a workaround; for me it looks like the way to go is to debug and fix a bug in QT itself. Of course, workaround would be easier, and it'd allow us to use older QT versions, but I ran out of ideas. Ideas are welcome :)

**Comment by kmike** [comment]
Based on a stacktrace, it looks like it is QtMIMETypeSniffer who tries to read the data from a closed reply. It adds a listener to `finished` signal here: https://github.com/qtproject/qtwebkit/blob/8b00fdada15a53c7764472435cffe04f22c3522f/Source/WebCore/platform/network/qt/QtMIMETypeSniffer.cpp#L39

**Comment by kmike** [comment]
but I don't have a clear picture of what's going on

**Comment by sunu** [comment]
Yeah, I saw it on the bug report. I guess it's gonna take a while now before the bug gets fixed and released. :disappointed:

**Comment by kmike** [comment]
Just tried rendering our mockserver with PhantomJS 2.0.0:

```
var url = "http://0.0.0.0:8998/slow.gif?n=5";

console.log('Loading a web page');
var page = require('webpage').create();
page.settings.resourceTimeout = 500;

page.open(url, function (status) {
  //Page is loaded!
  phantom.exit();
});
``` 

With resourceTimeout it hangs the same way as Splash (eating 100% CPU); without resourceTimeout it crashes.

**Comment by kmike** [comment]
My fix for https://bugreports.qt.io/browse/QTBUG-47654 is merged to qt and will be a part of 5.5.1 release. Then we can switch Docker image to use qt 5.5.1 and close this issue.

**Comment by kmike** [comment]
Docker image will use qt 5.5.1 in Splash 2.0.
